### PR TITLE
Bump repo to forc v0.35.3, fuels-rs v0.36.1, fuel-core v0.17.3, and rust v1.67.0

### LINKED ISSUE #96: Bump repo to forc v0.35.1
#### Description

Bump the repo to the latest version of forc

#### Motivation

Required for #95

### PR DESCRIPTION
**Pull request by bitzoic**
## Type of change

- Improvement (refactoring, restructuring repository, cleaning tech debt, ...)

## Changes

The following changes have been made:

- Updated libraries to forc v0.35.3
- Resolved warnings in the `UFP32` library
- Updated merkle proof to use new `From` and `.sha256()` additions to `Bytes` type
- Updated README on String that it uses `Bytes` and not `Vec`
- Updated test projects to account for stdlib changes
- Updated tests to fuel-rs v0.36.1
- Updated CI to new versions
- Updated README to new versions

## Notes

- `String` now does not support `append()`. It will be readded when https://github.com/FuelLabs/sway/issues/4158 is resolved. Added comments and documentation for this.
- Required for #95 

## Related Issues

Closes #96

### CODE REVIEW

File: libs/storagemapvec/src/lib.sw [hunk 1]
```diff
@@ -81,7 +81,14 @@ impl<K, V> StorageMapVec<K, V> {
...
         let key = sha256((key, len - 1, __get_storage_key()));
-        Option::Some::<V>(get::<V>(key))
+        match get::<V>(key) {
+            Option::Some(value) => {
+                Option::Some(value)
+            },
+            Option::None => {
+                Option::None
+            }
+        }
```
Comment (Braqzen) [inline libs/storagemapvec/src/lib.sw:91]: Can this be cleaner? Similarly line 130

File: tests/Forc.toml [hunk 2]
```diff
@@ -15,14 +15,12 @@ members = [
   "./src/unsigned_numbers/ufp32_div_test",
   "./src/unsigned_numbers/ufp32_exp_test",
   "./src/unsigned_numbers/ufp32_mul_test",
-  "./src/unsigned_numbers/ufp32_mul_test",
```
Comment (Braqzen) [inline tests/Forc.toml:18]: Removed 2 multiplication tests?

File: tests/src/merkle_proof/tests/utils/mod.rs [hunk 3]
```diff
@@ -2,13 +2,13 @@ use fuel_merkle::{
     binary::in_memory::MerkleTree,
     common::{empty_sum_sha256, Bytes32, LEAF, NODE},
 };
-use fuels::prelude::*;
+use fuels::{prelude::*, types::Bits256};
```
Comment (Braqzen) [inline tests/src/merkle_proof/tests/utils/mod.rs:5]: I'd make the `prelude::*` explicit while I'm at it

File: tests/src/ownership/tests/functions/owner.rs [hunk 4]
```diff
@@ -15,7 +15,7 @@ mod success {
 
         assert!(matches!(
             owner(&owner1.contract).await,
-            State::Uninitialized()
+            State::Uninitialized
```
Comment (Braqzen) [inline tests/src/ownership/tests/functions/owner.rs:18]: Does this work? Last time I used this the SDK team said that it gets treated as a function call so we must `call` it.

File: tests/src/signed_integers/signed_i128/tests/mod.rs [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
-
-        let _result = instance.main().call().await;
-    }
-}
+use fuels::prelude::*;
+
+abigen!(Script(
+    name = "Testi128",
+    abi = "src/signed_integers/signed_i128/out/debug/i128_test-abi.json"
+),);
```
Comment (Braqzen) [inline tests/src/signed_integers/signed_i128/tests/mod.rs:6]:
```suggestion
));
```

Formatter?

File: tests/src/signed_integers/signed_i128/tests/mod.rs [hunk 6]
```diff
@@ -1,21 +1,21 @@
...
+
+mod success {
+
+    use super::*;
+
+    #[tokio::test]
+    async fn runs_i128_test_script() {
+        let path_to_bin = "src/signed_integers/signed_i128/out/debug/i128_test.bin";
```
Comment (Braqzen) [inline tests/src/signed_integers/signed_i128/tests/mod.rs:14]: Perhaps factor out into module?

File: tests/src/string/src/main.sw [hunk 7]
```diff
@@ -91,10 +111,10 @@ impl StringTest for Contract {
         assert(string.capacity() == 16);
 
         string.clear();
-        assert(string.capacity() == 16);
+        assert(string.capacity() == 0);
 
         string.push(NUMBER0);
-        assert(string.capacity() == 16);
+        assert(string.capacity() == 1);
```
Comment (Braqzen) [inline tests/src/string/src/main.sw:117]: I guess these weren't working before

File: libs/fixed_point/src/ufp32.sw [hunk 8]
```diff
@@ -98,7 +98,7 @@ impl core::ops::Multiply for UFP32 {
         }
 
         Self {
-            value: res_u64,
+            value: asm(ptr: res_u64) { ptr: u32 },
```
Comment (Braqzen) [inline libs/fixed_point/src/ufp32.sw:101]: Why assembly?

File: libs/fixed_point/src/ufp32.sw [hunk 9]
```diff
@@ -250,7 +250,7 @@ impl Exponentiate for UFP32 {
         // which means that the denominator is always 2 ^ 16
         // we need to delete the nominator by 2 ^ (16 * exponent - 1)
         // - 1 is the formula is due to denominator need to stay 2 ^ 16
-        let nominator = nominator_pow >> demoninator_power * (exponent_int - 1);
+        let nominator = nominator_pow >> demoninator_power * (exponent_int - 1_u32);
```
Comment (Braqzen) [inline libs/fixed_point/src/ufp32.sw:253]: Does this change affect anything?

File: tests/Forc.toml (hunk 2 above)
Comment (bitzoic) [reply tests/Forc.toml:18]: They are duplicates from the line above

File: tests/src/ownership/tests/functions/owner.rs (hunk 4 above)
Comment (bitzoic) [reply tests/src/ownership/tests/functions/owner.rs:18]: `State::Uninitialized()` results in an error

File: tests/src/signed_integers/signed_i128/tests/mod.rs (hunk 5 above)
Comment (bitzoic) [reply tests/src/signed_integers/signed_i128/tests/mod.rs:6]: Yeah `cargo fmt`

File: tests/src/string/src/main.sw (hunk 7 above)
Comment (bitzoic) [reply tests/src/string/src/main.sw:117]: No, the `clear()` function in the stdlib has changed to modify capacity.

File: libs/fixed_point/src/ufp32.sw (hunk 8 above)
Comment (bitzoic) [reply libs/fixed_point/src/ufp32.sw:101]: Remove warnings of converting `u64` to `u32`

File: libs/fixed_point/src/ufp32.sw (hunk 9 above)
Comment (bitzoic) [reply libs/fixed_point/src/ufp32.sw:253]: Removing annoying warnings

File: libs/storagemapvec/src/lib.sw (hunk 1 above)
Comment (bitzoic) [reply libs/storagemapvec/src/lib.sw:91]: tbh not sure what I was thinking here. Updated to just `get::<V>(key)`

File: tests/src/signed_integers/signed_i128/tests/mod.rs (hunk 6 above)
Comment (bitzoic) [reply tests/src/signed_integers/signed_i128/tests/mod.rs:14]: The single line? Not really sure what you mean by this

Comment (Braqzen) [reply tests/src/signed_integers/signed_i128/tests/mod.rs:14]: Lots of hardcoded strings for binaries across files. I would consider whether we should factor out

File: libs/nft/src/extensions/burnable/burnable.sw [hunk 10]
```diff
@@ -26,8 +26,8 @@ impl Burnable for NFTCore {
     fn burn(self) {
         require(self.owner == msg_sender().unwrap(), AccessError::SenderNotOwner);
 
-        store(sha256((BALANCES, self.owner)), get::<u64>(sha256((BALANCES, self.owner))) - 1);
-        store(sha256((TOKENS, self.token_id)), Option::None::<NFTCore>());
+        store(sha256((BALANCES, self.owner)), get::<u64>(sha256((BALANCES, self.owner))).unwrap() - 1);
```
Comment (matt-user) [inline libs/nft/src/extensions/burnable/burnable.sw:29]:
```suggestion
        store(sha256((BALANCES, self.owner)), get::<u64>(sha256((BALANCES, self.owner))).unwrap_or(0) - 1);
```
thoughts?

File: libs/signed_integers/src/i128.sw [hunk 11]
```diff
@@ -95,9 +95,7 @@ impl core::ops::Add for I128 {
     fn add(self, other: Self) -> Self {
         // subtract 1 << 63 to avoid double move
         let mut res = Self::new();
-        if (self.underlying > Self::indent()
-            || self.underlying == Self::indent())
-        {
+        if (self.underlying > Self::indent() || self.underlying == Self::indent()) {
```
Comment (matt-user) [inline libs/signed_integers/src/i128.sw:98]:
```suggestion
        if (self.underlying >= Self::indent()) {
```

File: libs/signed_integers/src/i256.sw [hunk 12]
```diff
@@ -96,9 +96,7 @@ impl core::ops::Add for I256 {
     fn add(self, other: Self) -> Self {
         // subtract 1 << 63 to avoid double move
         let mut res = Self::new();
-        if (self.underlying > Self::indent()
-            || self.underlying == Self::indent())
-        {
+        if (self.underlying > Self::indent() || self.underlying == Self::indent()) {
```
Comment (matt-user) [inline libs/signed_integers/src/i256.sw:99]:
```suggestion
        if (self.underlying >= Self::indent()) {
```

File: libs/signed_integers/src/i256.sw [hunk 13]
```diff
@@ -174,11 +172,7 @@ impl core::ops::Subtract for I256 {
     /// Subtract a I256 from a I256. Panics of overflow.
     fn subtract(self, other: Self) -> Self {
         let mut res = Self::new();
-        if (self.underlying > Self::indent()
-            || self.underlying == Self::indent())
-            && (other.underlying > Self::indent()
-            || other.underlying == Self::indent())
-        {
+        if (self.underlying > Self::indent() || self.underlying == Self::indent()) && (other.underlying > Self::indent() || other.underlying == Self::indent()) {
```
Comment (matt-user) [inline libs/signed_integers/src/i256.sw:175]:
```suggestion
        if (self.underlying >= Self::indent() && (other.underlying >= Self::indent()) {
```

**Comment by Braqzen** [comment]
[This](https://github.com/FuelLabs/sway/issues/4125) bug needs to be fixed first

**Comment by Braqzen** [comment]
I recommend `forc 0.35.2` found via [this channel](https://github.com/FuelLabs/fuelup/blob/gh-pages/channels/latest/channel-fuel-latest-2023-02-18.toml)

File: libs/signed_integers/src/i128.sw (hunk 11 above)
Comment (bitzoic) [reply libs/signed_integers/src/i128.sw:98]: I forgot why this was done exactly but I know it was on purpose. @rostyslavtyshko should be able to explain

File: libs/nft/src/extensions/burnable/burnable.sw (hunk 10 above)
Comment (bitzoic) [reply libs/nft/src/extensions/burnable/burnable.sw:29]: 0 - 1 would revert due to a negative number. This value shouldn't ever be zero if we reach this far into the `burn()` function due to the revert case above.

**Comment by Braqzen** [comment]
> [This](https://github.com/FuelLabs/sway/issues/4125) bug needs to be fixed first

Once a release is made bump to `forc 0.35.3` instead.

File: libs/signed_integers/src/i128.sw (hunk 11 above)
Comment (matt-user) [reply libs/signed_integers/src/i128.sw:98]: OK I'll mark this as resolved then

**Comment by nfurfaro** [comment]
Looks like forc 0.35.3 is now released.

**Comment by Braqzen** [comment]
> Looks like forc 0.35.3 is now released.

Need to wait for the binaries to be made available and if we use version pinning here then we also need to wait till fuelup creates a new file to pull these changes from

**Review by matt-user — APPROVED** [review]
lgtm

**Review by Braqzen — APPROVED** [review]

**Pull request merged**